prov/verbs/src/verbs_info.c

```diff
@@ -834,53 +826,67 @@ static int fi_ibv_alloc_info(struct ibv_context *ctx, struct fi_info **info,
 	return ret;
 }
 
-static int fi_ibv_copy_ifaddr(const char *name, const char *service, uint64_t flags,
-		struct fi_info *info)
+static void fi_ibv_verbs_devs_free(struct dlist_entry *verbs_devs)
 {
-	struct rdma_addrinfo *rai;
-	struct fi_info *fi;
-	struct rdma_cm_id *id;
-	int ret;
-
-	ret = fi_ibv_get_rdma_rai(name, service, flags, NULL, &rai);
-	if (ret) {
-		FI_WARN(&fi_ibv_prov, FI_LOG_FABRIC,
-				"rdma_getaddrinfo failed for name:%s\n", name);
-		return ret;
-	}
-	ret = rdma_create_ep(&id, rai, NULL, NULL);
-	if (!ret) {
-		for (fi = info; fi; fi = fi->next)
-			if (!strncmp(id->verbs->device->name, fi->domain_attr->name,
-						strlen(id->verbs->device->name)))
-				break;
-		if (!fi) {
-			FI_WARN(&fi_ibv_prov, FI_LOG_FABRIC,
-					"No matching fi_info for device: "
-					"%s with address: %s\n",
-					id->verbs->device->name, name);
-		} else {
-			if (fi->src_addr) {
-				free(fi->src_addr);
-				fi->src_addr = NULL;
-			}
-			fi_ibv_rai_to_fi(rai, fi);
+	struct verbs_dev_info *dev;
+	struct verbs_addr *addr;
+
+	while (!dlist_empty(verbs_devs)) {
+		dlist_remove_container(verbs_devs->next, dev, entry);
+		while (!dlist_empty(&dev->addrs)) {
+			dlist_remove_container(dev->addrs.next, addr, entry);
+			rdma_freeaddrinfo(addr->rai);
+			free(addr);
 		}
-		rdma_destroy_ep(id);
+		free(dev->name);
+		free(dev);
+	}
+}
+
+static int fi_ibv_add_rai(struct dlist_entry *verbs_devs, struct rdma_cm_id *id,
+		struct rdma_addrinfo *rai)
+{
+	struct verbs_dev_info *dev;
+	struct verbs_addr *addr;
+	const char *dev_name;
+
+	dev_name = ibv_get_device_name(id->verbs->device);
+	dlist_foreach_container(verbs_devs, dev, entry)
+		if (!strcmp(dev_name, dev->name))
+			goto add_rai;
+
+	if (!(dev = malloc(sizeof(*dev))))
+		return -FI_ENOMEM;
+
+	if (!(dev->name = strdup(dev_name))) {
+		free(dev);
+		return -FI_ENOMEM;
 	}
-	rdma_freeaddrinfo(rai);
+
+	dlist_init(&dev->addrs);
+	dlist_insert_tail(&dev->entry, verbs_devs);
+
+add_rai:
+	if (!(addr = malloc(sizeof(*addr))))
+		return -FI_ENOMEM;
```

Comment:
The function which calls fi_ibv_add_rai on receiving an error would clean up the entire list. So I dropped the cleanup here.